### ISSUE TITLE
Compatibilty with Nextcloud 21

### ISSUE THREAD

**Issue by christianlupus** [nextcloud/cookbook#341]
The current `master` branch on the NC server repo is of version 21. Or app is not declared to be compatible with version 21 yet. Checking is needed.

**Comment by IzzySoft** [comment]
Speaking of which: could you please point out what are the resp. last versions compatible with PHP 7.0, PHP 7.1? I just now HEAD requires 7.2 (as that's where the dependency was added to `appinfo/info.xml`).

Background of the question: some folks (like me) are still stuck on some older PHP version because those are what their LTS (here: Linux Mint 18.3 with PHP 7.0) ship, and upgrading to the next LTS is not always a (short-term) option. Currently, Nextcloud offers me to upgrade to 0.7.6, and that would certainly break Cookbook on my install (side-mark: at least NC 15 seems to ignore the php minVer field in `info.xml` even if it's there; IIRC there was already an issue for that "over there").

Maybe a short note in the FAQ section of the Readme could be added? For now, that would only be 2 entries which could go into a single line: "xx is the last version supporting PHP 7.0, yy for PHP 7.1"). I'd really like to update from my current 0.6.3, but I don't dare…

**Comment by christianlupus** [comment]
> Speaking of which: could you please point out what are the resp. last versions compatible with PHP 7.0, PHP 7.1? I just now HEAD requires 7.2 (as that's where the dependency was added to `appinfo/info.xml`).

Unfortunately, I do not know. I just found your issue #282. Maybe it's worth running with an older version in a dev environment. If it is only a small fix that might bring the old PHP version (at least the 7.x) back, why not.

> Background of the question: some folks (like me) are still stuck on some older PHP version because those are what their LTS (here: Linux Mint 18.3 with PHP 7.0) ship, and upgrading to the next LTS is not always a (short-term) option. Currently, Nextcloud offers me to upgrade to 0.7.6, and that would certainly break Cookbook on my install (side-mark: at least NC 15 seems to ignore the php minVer field in `info.xml` even if it's there; IIRC there was already an issue for that "over there").

I am perfectly sure that current installations are not ignoring the settings. During my work for #337 I found out the hard way, that PHP 7.1 was refused by the installer

> Maybe a short note in the FAQ section of the Readme could be added? For now, that would only be 2 entries which could go into a single line: "xx is the last version supporting PHP 7.0, yy for PHP 7.1"). I'd really like to update from my current 0.6.3, but I don't dare…

If you find the time to test it out, you are welcome to share your knowledge. I will happily add it to the docs/readme. Unfortunatelyly, my own time is a bit restricted these days.

**Comment by IzzySoft** [comment]
> I just found your issue #282

I knew I had asked before – but couldn't find where. You were more successfull finding it, thanks! Gives me a more complete picture again. Especially it saves me the headache figuring how to fix it again when it breaks.

> Maybe it's worth running with an older version in a dev environment.

I might try that on one of my installations – if I find one where no other apps have updates pending (found out the hard way recently NC then tries to update everything, thus breaking even more).

> If it is only a small fix that might bring the old PHP version (at least the 7.x) back, why not.

Yes, I fully agree PHP < 7 can be ignored; I'm not aware of any still supported distri shipping with such.

> I am perfectly sure that current installations are not ignoring the settings.

That's great to read! You don't happen to remember which version you experienced that with? Must be NC < 21, when I look at the other issues in the tracker here. But PHP 7.1? NC 15 is the last one supporting 7.0, NC 16 already requires 7.1+ (or I had long updated my installs).

> If you find the time to test it out, you are welcome to share your knowledge.

Will do: just found one instance where only a single other app is signaling updates (and definitely would break then). It's non-essential on that install, so I disabled it – and successfully installed Cookbook 0.6.4. Now let me be adventurous, so I can at least fill the PHP 7.0 gap. Will add another comment when done, as I must postpone the test a little (other tasks just called me, personally :rofl:)

Thanks!

**Comment by IzzySoft** [comment]
OK, that was a short test: the last one supporting PHP 7.0 seems to be Cookbook 0.6.4. If I try replacing that by 0.6.5, Nextcloud wants to start the big update cycle – meaning it insists on updating to the very latest version from the store. Which of course breaks: opening the app results in a 500 server error.

```
[index] Error: OCP\AppFramework\QueryException: Could not resolve OCA\Cookbook\Controller\MainController! Class OCA\Cookbook\Service\RecipeService does not exist at <<closure>>

0. /usr/local/share/nextcloud/lib/private/AppFramework/Utility/SimpleContainer.php line 119
   resolve("OCA\\Cookbook\\ ... r")
1. /usr/local/share/nextcloud/lib/private/AppFramework/DependencyInjection/DIContainer.php line 426
   query("OCA\\Cookbook\\ ... r")
2. /usr/local/share/nextcloud/lib/private/AppFramework/DependencyInjection/DIContainer.php line 397
   queryNoFallback("OCA\\Cookbook\\ ... r")
3. /usr/local/share/nextcloud/lib/private/AppFramework/App.php line 105
   query("OCA\\Cookbook\\ ... r")
4. /usr/local/share/nextcloud/lib/private/AppFramework/Routing/RouteActionHandler.php line 47
   main("OCA\\Cookbook\\ ... r", "index", OC\AppFramework\ ... {}, {_route: "cookbook.main.index"})
5. <<closure>>
   __invoke({_route: "cookbook.main.index"})
6. /usr/local/share/nextcloud/lib/private/Route/Router.php line 297
   call_user_func(OC\AppFramework\ ... {}, {_route: "cookbook.main.index"})
7. /usr/local/share/nextcloud/lib/base.php line 987
   match("/apps/cookbook/")
8. /usr/local/share/nextcloud/index.php line 42
   handleRequest()

GET /nextcloud/index.php/apps/cookbook/
```

Does Cookbook have any database tables I now need to clean up, to remove it entirely from that install – e.g. to try a clean install?

**Comment by IzzySoft** [comment]
Ah, just noted from the other issue:

> I know 0.6.5 does not work with PHP7.0, but skipped 0.6.4 when "running down"

So the first fact you could point out in the Readme would then be: "last version running with PHP 7.0 is v0.6.4". But then: Trying to import recipes fails (Javascript popup: "recipe could not be added. Please activate Javascript"). Needless to say that Javascript is active. Downgraded to 0.6.3, truncated all cookbook tables, removed the JSON of the failed import – and import works again.

**TL;DR:** v0.6.3 is the latest working version for PHP 7.0, unfortunately. For PHP 7.1, someone else needs to test – I don't have a matching environment.

**Comment by christianlupus** [comment]
> That's great to read! You don't happen to remember which version you experienced that with? Must be NC < 21, when I look at the other issues in the tracker here. But PHP 7.1? NC 15 is the last one supporting 7.0, NC 16 already requires 7.1+ (or I had long updated my installs).

I used the stable19 branch of NC with a docker container for PHP 7.1 and the current master branch of the cookbook app.

> Does Cookbook have any database tables I now need to clean up, to remove it entirely from that install – e.g. to try a clean install?

There are a few tables in the DB that are from the cookbook app (should be around three). They are all named with a `oc_cookbook_` prefix.

I will have a look once in a while. Unfortunately, I have some personal pressing issues that allow me to invest only a limited amount of time into this app at the moment.

**Comment by IzzySoft** [comment]
> I used the stable19 branch of NC with a docker container for PHP 7.1 and the current master branch of the cookbook app.

Ah, OK – thanks! So at least NC 19+ honor that field.

> There are a few tables in the DB

Yes, identified the 3. Truncating them allowed me to start over and cross-verify that the "could not import" issue wasn't due to downgrading.

And thanks for caring! No pressure, I'll keep patient. After all, I should really upgrade my system (once I get everything sorted out), which would solve this issue as well for me…

**Comment by obbardc** [comment]
And don't forget nextcloud v20 ;-)

**Comment by christianlupus** [comment]
v20 is handled in #331.

**Comment by christianlupus** [comment]
The new version 0.8.3 was just pushed out with support for NC 21.
